ref(discover): Remove all references to QueryFilter

shruthilayaj opened the pull request:
Remove all references to QueryFilter, QueryFields and
QueryBase. Introduced a new param to QueryBuilder
called `partial` that only partially initializes QueryBuilder
so that the Filter and Fields code can be tested in isolation.
Open to other suggestions on how to do that.

Moved groupby from being a class property to class
attribute everywhere (this was a bit inconsistent in places)

shruthilayaj committed:
check in

shruthilayaj committed:
check in progress

shruthilayaj committed:
check in progress

shruthilayaj committed:
check in progress

shruthilayaj committed:
Merge branch 'master' into feat/add-data-selector-to-query-builder

shruthilayaj committed:
revert changes to base, filter and fields

shruthilayaj committed:
clean up dataset config

shruthilayaj committed:
refactor

shruthilayaj committed:
error if not dataset

shruthilayaj committed:
update trend query

shruthilayaj committed:
mock stuff

shruthilayaj committed:
more type checks

shruthilayaj committed:
fix imports

shruthilayaj committed:
move resolve division to Query Builder

shruthilayaj committed:
rearrange functions

shruthilayaj committed:
don't use deprecared decorators

shruthilayaj committed:
feat(discover): TimeseriesQueryBuilder inherits from QueryBuilder

Updating remaing QueryBuilder classes so they don't inherit
or instatiate QueryFilter.

shruthilayaj committed:
check in progress

shruthilayaj committed:
check in progress

shruthilayaj committed:
merge with master

shruthilayaj committed:
check in progress

shruthilayaj committed:
remove references to QueryField and QueryBase

shruthilayaj committed:
remove references to query fields

shruthilayaj committed:
make group by an attribute instead of prop

shruthilayaj committed:
fix more tests

shruthilayaj committed:
fix last bit of tests

Zylphrex commented on src/sentry/search/events/builder.py:
```diff
@@ -111,14 +115,18 @@ def __init__(
-        self.array_join = None if array_join is None else [self.resolve_column(array_join)]
+
+        # TODO(wmak): Refactor snuba/metrics.py to call QueryBuilder correctly.
+        if not partial:
```
Instead of a `partial` param, what if the constructor doesn't do any of the resolutions automatically? Instead, we can move this into a helper like

```python
def resolve_everything(self):
  # do the resolution here can
  # this can be overridden by child classes
  # (i think trends needs this) to change up how the resolution works
```
Then the `run_query` proposed in #31354 can just call it.

This way, you don't need to special case the uses where they just want to use the builder to resolve a few things and we still get the same convenient interface to work with. Thoughts?

shruthilayaj replied:
Yeah I had the same idea and talked about it with @wmak and he brought up a good point that it's kinda nice to be able resolve it in the constructor because it gives the caller the flexibility to add to or remove from these attributes (looks like the trends functions do that because we resolve the trends columns outside the query builder [here](https://github.com/getsentry/sentry/blob/ac6daa30918636b2a38bea735e7158e96f876e1c/src/sentry/api/endpoints/organization_events_trends.py#L477-L489))

shruthilayaj replied:
But yeah this partial thing kinda feels hacky

Zylphrex replied:
I'm not sure I agree with it being nice that you are able to add/remove these attributes outside of the query builder. I'm also guilty of doing this in the suspect spans query where I append additional conditions. While convenient, it moves some of the responsibility of building the query outside of the query builder. It also leads to hacks like that in the trends query builder where it doesn't specify an order by in the constructor only to manually add one later on.

In this world where we have this `resolve_everything` call, we are able to subclass the query builder and implement a custom query resolution that can include adding/removing from the query attributes.

Zylphrex replied:
I like the idea of another class with an empty `resolve_everything`. If the class is named appropriately, this would cater to both use cases in a clean way.

shruthilayaj replied:
Oooh yeah maybe let me try it out with a "resolve_query" method

wmak replied:
oh yeah `resolve_query` is a far better method name :+1:

untitaker commented:
@shruthilayaj the python-typing "team" was roped in for code review because of a change to mypy.ini. Are you looking for substantial code review wrt Python typing here or is this a fluke? On the chance it's the former could you point out files you want review on, because this PR is very large. Thanks!

shruthilayaj commented:
> 

@untitaker ah no review required from python typing - I think it auto requested because I changed the `mypy.ini` file but that was because I deleted a file and removed it from type checks. This is just a discover query builder refactor!

shruthilayaj committed:
Move resolving the query to a parent class

shruthilayaj committed:
type checks

Zylphrex commented on src/sentry/search/events/builder.py:
```diff
@@ -60,28 +60,21 @@
 
 
-class QueryBuilder:
+class QueryComplier:
```
Do you mean `QueryCompiler`?

shruthilayaj replied:
yes, 100%

shruthilayaj committed:
typo

wmak reviewed (approved):
<img width="523" alt="image" src="https://user-images.githubusercontent.com/4205004/151453563-7f44b78f-6e2b-4f27-a787-7546f50f4278.png">
🔥 beautiful

shruthilayaj committed:
Merge branch 'master' into feat/move-remaining-query-builders-to-use-config

shruthilayaj committed:
have unresolved query just override init

shruthilayaj committed:
simplify by removing resolve query

shruthilayaj committed:
simplify by removing resolve query

shruthilayaj committed:
override resolve_query

shruthilayaj committed:
fix typing errors
